src/ssl/gadgets.cc

```diff
@@ -9,35 +9,28 @@
 #include "squid.h"
 #include "ssl/gadgets.h"
 
-EVP_PKEY * Ssl::createSslPrivateKey()
+static EVP_PKEY *
+CreateRsaPrivateKey()
 {
-    Security::PrivateKeyPointer pkey(EVP_PKEY_new());
-
-    if (!pkey)
-        return NULL;
-
-    BIGNUM_Pointer bn(BN_new());
-    if (!bn)
-        return NULL;
-
-    if (!BN_set_word(bn.get(), RSA_F4))
-        return NULL;
-
-    Ssl::RSA_Pointer rsa(RSA_new());
+    Ssl::EVP_PKEY_CTX_Pointer rsa(EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, nullptr));
     if (!rsa)
-        return NULL;
+        return nullptr;
 
-    int num = 2048; // Maybe use 4096 RSA keys, or better make it configurable?
-    if (!RSA_generate_key_ex(rsa.get(), num, bn.get(), NULL))
-        return NULL;
+    if (EVP_PKEY_keygen_init(rsa.get()) <= 0)
+        return nullptr;
 
-    if (!rsa)
-        return NULL;
+    int num = 2048; // Maybe use 4096 RSA keys, or better make it configurable?
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(rsa.get(), num) <= 0)
+        return nullptr;
 
-    if (!EVP_PKEY_assign_RSA(pkey.get(), (rsa.get())))
-        return NULL;
+    /* Generate key */
+    Security::PrivateKeyPointer pkey(EVP_PKEY_new());
+    if (pkey) {
+        auto *foo = pkey.get();
+        if (EVP_PKEY_keygen(rsa.get(), &foo) <= 0)
+            return nullptr;
+    }
```

Comment:
I do not think that this part of code can work. Without studying a lot the related OpenSSL api my sense is that you need something like the following:

  `EVP_PKEY *pkey = NULL;`
  `EVP_PKEY_keygen(rsa.get(), pkey);`
  `return pkey;`

Reply:
Thanks. Done that.